### PR TITLE
Add Primary Key to CDK.

### LINKED ISSUE #2766: santa: primary keys
## Tell us about the problem you're trying to solve
Allow defining primary keys at the stream level and reflecting them in the catalog

### PR DESCRIPTION
**Pull request by davinchia**
## What
Allow CDK users to set primary keys in the Streams.

Closes: #2766.

## How
* Make primary key an abstract internal method, `_get_source_defined_primary_keys`, in the base `stream` class. Modify the `as_airbyte_stream` function to use the abstract method when setting `source_defined_primary_keys`.
* Create a `source_defined_primary_keys` property on `HttpStream` that is a list of strings. Implement the abstract primary keys method in `HttpStream` converting the list of keys to a list of list of keys. I felt reducing the interface to a list of strings allows users to avoid the confusion of a list of list of strings and only thinking about passing in all the response' fields required for the primary key.

## Recommended reading order
1. `core.py`
2. `http.py`

### CODE REVIEW

File: airbyte-integrations/bases/base-python/base_python/cdk/streams/core.py
Comment (sherifnada) [inline airbyte-integrations/bases/base-python/base_python/cdk/streams/core.py:70]: pk is optional
```suggestion
    ) -> Optional[List[List[str]]]: 
```

**Review by sherifnada — CHANGES_REQUESTED** [review]
bunch of suggestions and one question

File: airbyte-integrations/bases/base-python/base_python/cdk/streams/core.py
Comment (davinchia) [reply airbyte-integrations/bases/base-python/base_python/cdk/streams/core.py:68]: Spoke offline.

Agreed that doing this provides sane default while exposing all underlying functionality.

**Comment by davinchia** [comment]
How does this look @sherifnada ?

File: airbyte-integrations/bases/base-python/base_python/catalog_helpers.py
Comment (sherifnada) [inline airbyte-integrations/bases/base-python/base_python/catalog_helpers.py:35]: this is a named field in the airbyte protocol -- `AirbyteStream.cursor` doesn't exist

File: airbyte-integrations/bases/base-python/base_python/cdk/streams/core.py
Comment (sherifnada) [inline airbyte-integrations/bases/base-python/base_python/cdk/streams/core.py:110]: can either keep this as `source_defined_cursor`? this one makes sense because the user actually can input a cursor, and this just toggles that

File: airbyte-integrations/bases/base-python/base_python/cdk/streams/core.py
Comment (sherifnada) [inline airbyte-integrations/bases/base-python/base_python/cdk/streams/core.py:120]:
```suggestion
        :return: string if single primary key, list of strings if composite primary key, list of list of strings if composite primary key consisting of nested fields.
        If the stream has no primary keys, return None.
```

File: airbyte-integrations/bases/base-python/base_python/cdk/streams/http.py
Comment (sherifnada) [inline airbyte-integrations/bases/base-python/base_python/cdk/streams/http.py:41]: why set this to an empty string? I think either we make the base property not abstract and return None, or require users to choose a primary key. I'm kind of on the side of making it abstract and not overriding it here because almost all data has a primary key, so simply asking the user about the primary key will result in a much higher rate of connectors declaring primary keys.

**Review by sherifnada — APPROVED** [review]
small comments and one breaking change that will need to be removed but LGTM

File: airbyte-integrations/bases/base-python/base_python/catalog_helpers.py
Comment (davinchia) [reply airbyte-integrations/bases/base-python/base_python/catalog_helpers.py:35]: whoops. accidental refactor.

File: airbyte-integrations/bases/base-python/base_python/cdk/streams/http.py
Comment (davinchia) [reply airbyte-integrations/bases/base-python/base_python/cdk/streams/http.py:41]: good point.

**Comment by davinchia** [comment]
@sherifnada holding off on publishing this base python version and updating all the down stream consumers since we are refactoring this. will do so in a separate PR if we need to do so before the refactor goes out.

**Pull request merged**
